### ISSUE TITLE
Footgun with Rc::assume_init and related methods

### ISSUE THREAD

**Issue by Diggsey** [rust-lang/rust#131861]
### Location

I came across https://doc.rust-lang.org/nightly/std/rc/struct.Rc.html#method.assume_init recently, and spotted a footgun that I think ought to be called out:

### Summary

The safety section does not clarify whether multiple `Rc`s are allowed to exist when `assume_init` is called. If they are, then whether `Drop` is called on the inner value will depend on the drop order of those `Rc`s. In the case of `Arc` this might well be non-deterministic.

IMO, the documentation should specify whether this is allowed, and if so should point out that callers must take care around this potential issue.

**Comment by lukaslueg** [comment]
If we `Rc::<MaybeUninit<T>>::assume_init()` one instance, other instances remain in the `Rc<MaybeUninit<T>>` state, and those will never call `drop` on `T`, unless those instances assume initialization as well. Getting this right - if one instance can assume initialization, everyone has to - is left as an exercise to the reader. FWIW I fail to see how this could lead to memory unsafety. But it should be fairly easy to leak memory/resources that way

* Initialize the inner `T` where `T: Drop`
* At least one instance of `Rc<MaybeUninit<T>>` remains uninitialized even when `T` is in fact initialized
* The above instance will prevent other `Rc<T>` from dropping the inner `T` when they go out of scope
* The remaining `Rc<MaybeUninit<T>>` does not drop `T` because `MaybeUninit<T>` never does.
* `T` remains alive forever and is unreachable

**Comment by Diggsey** [comment]
> FWIW I fail to see how this could lead to memory unsafety.

eg. `Pin` requires that the destructor is called before the memory is reused, but `Rc<MaybeUninit<T>>` will free the memory without calling the destructor.

**Comment by hanna-kruppe** [comment]
That's a good point. Luckily, there are some factors that work against this happening by accident:

1. The natural and recommended way to do initialization goes through `&mut MaybeUninit<T>` to get a `*mut T` to write through. Getting `&mut T` from `Rc<T>` or `Arc<T>` requires exclusive access, i.e., refcount == 1, so there can't be other references to the same `MaybeUninit<T>` floating around.
2. Typically you'd call `assume_init` immediately after doing the initialization, before doing anything that might create more references.

Of course, it would still be good to put this non-obvious gotchas, especially around `Pin`, in the docs. Also, I think you can technically do the initialization without UB using just shared references and pointer casts (i.e., without checking refcount==1) if you involve `UnsafeCell`, but that seems very convoluted and I don't know of any case where it's useful or easier than the `&mut MaybeUninit<T>` route.

**Comment by Diggsey** [comment]
Right - I think the std library should be opinionated here - ie. either say that it's an error to call `assume_init` when multiple references exist (and verify this in debug) or say it's allowed but document the behaviour (that last Rc to drop determines drop behaviour)

**Comment by workingjubilee** [comment]
@joshtriplett was this ever discussed before it was FCPed?

**Comment by joshtriplett** [comment]
> [@joshtriplett](https://github.com/joshtriplett) was this ever discussed before it was FCPed?

I don't think this issue ever came up.

**Comment by workingjubilee** [comment]
@rust-lang/libs-api This nomination is because libs-api FCPed a number of `unsafe fn` constructors on a type that can have a destructor. Leaving the contract with regards to the destructor unspecified makes this an avenue for unsound behavior in unsafe code. Technically this is no worse than simply cloning an Rc and then calling `mem::forget` on it, but it seems undesirable to add such avenues.

**Comment by cuviper** [comment]
> either say that it's an error to call `assume_init` when multiple references exist (and verify this in debug)

I don't think this should be an error, because you might be in the middle of calling `assume_init` on all of them. e.g. a graph could prepare shared node pointers this way, and then `assume_init` everything into a final type.

> or say it's allowed but document the behaviour (that last Rc to drop determines drop behaviour)

So this is my preference.

**Comment by Diggsey** [comment]
Another option would be to say that when any `Rc` or `Arc` is dropped or cloned, all of the related pointers must be pointing to the same type, or else it's UB. (In order to avoid being too prescriptive on the implementation)

**Comment by joshtriplett** [comment]
I think we should specify "last Rc/Arc dropped determines drop behavior". It's no more unsafe than `mem::forget`, it's just a footgun that people should be aware of when calling the unsafe method.

**Comment by Amanieu** [comment]
We discussed this in the libs-api meeting today. We agreed that it would be best to document the current behavior ("last Rc/Arc dropped determines drop behavior"), but we should also include a strong recommendation to only do this when there is only a single active reference. It would be difficult to use this without `get_mut` anyways since you would need interior mutability *within* the `MaybeUninit` which already requires quite a lot of unsafe code to work.

Once `UniqueRc` is stabilized then we should consider deprecating this method in favor of a similar method on `UniqueRc`.
